docs/tutorials/core-from-scratch.ipynb

```diff
@@ -0,0 +1,369 @@
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "12",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from jax.interpreters import ad\n",
+    "\n",
+    "\n",
+    "@jax.custom_jvp\n",
+    "def kepler_solver(mean_anom, ecc):\n",
+    "    return kepler_solver_impl(mean_anom, ecc)\n",
+    "\n",
+    "\n",
+    "@kepler_solver.defjvp\n",
+    "def kepler_solver_jvp(primals, tangents):\n",
+    "    mean_anom, ecc = primals\n",
+    "    d_mean_anom, d_ecc = tangents\n",
+    "\n",
+    "    # Run the solver from above to compute `ecc_anom`\n",
+    "    ecc_anom = kepler_solver(mean_anom, ecc)\n",
+    "\n",
+    "    # Propagate the derivatives using the implicit function theorem\n",
+    "    dEdM = 1 / (1 - ecc * jnp.cos(ecc_anom))\n",
+    "    dEde = jnp.sin(ecc_anom) * dEdM\n",
+    "    d_ecc_anom = dEdM * make_zero(d_mean_anom) + dEde * make_zero(d_ecc)\n",
+    "\n",
+    "    return ecc_anom, d_ecc_anom\n",
+    "\n",
+    "\n",
+    "def make_zero(tan):\n",
+    "    # This is a helper function to handle symbolic zeros (i.e. parameters\n",
+    "    # that are not being differentiated)\n",
+    "    if type(tan) is ad.Zero:\n",
+    "        return ad.zeros_like_aval(tan.aval)\n",
+    "    else:\n",
+    "        return tan"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "13",
+   "metadata": {},
+   "source": [
+    "Given this final `kepler_solver` function, we can take a look at the derivatives of the eccentric anomaly as a function of mean anomaly and eccentricity.\n",
+    "By default JAX only supports differentiation of scalar functions, but combining `jax.grad` with `jax.vmap` we can get efficient computation of the elementwise derivatives of this function."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "14",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def elementwise_grad(argnums):\n",
+    "    return jax.vmap(jax.grad(kepler_solver, argnums=argnums), in_axes=(0, None))\n",
+    "\n",
+    "\n",
+    "fig, axes = plt.subplots(3, 1, sharex=True)\n",
+    "\n",
+    "ax = axes[0]\n",
+    "ax.plot(mean_anom, kepler_solver(mean_anom, ecc), \"k\")\n",
+    "ax.set_ylabel(\"E\")\n",
+    "\n",
+    "ax = axes[1]\n",
+    "ax.plot(mean_anom, elementwise_grad(0)(mean_anom, ecc), \"k\")\n",
+    "ax.set_ylabel(\"dE / dM\")\n",
+    "\n",
+    "ax = axes[2]\n",
+    "ax.plot(mean_anom, elementwise_grad(1)(mean_anom, ecc), \"k\")\n",
+    "ax.set_xlabel(\"mean anomaly; M\")\n",
+    "ax.set_ylabel(\"dE / de\")\n",
+    "ax.set_xlim(0, 2 * jnp.pi);"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "15",
+   "metadata": {},
+   "source": [
+    "Finally, we can test the performance of this operation to see that the computational cost is still dominated by the solve.\n",
+    "In other words, the total cost of the of computing these derivatives is only fractionally (about 1% on my laptop) higher than just solving the equation itself."
```

Comment:
Perhaps add "In comparison, the runtime of the default traced gradient computation is higher by a factor of four."

Reply:
Added.